### ISSUE TITLE
Start-IcingaAgentInstallWizard always brings "failed to connect to director"

### ISSUE THREAD

**Issue by blindzero** [Icinga/icinga-powershell-framework#59]
I have a running Icinga director 1.7.2 setup with several (old-school) Icinga2 for Windows agents running. The old director kickstart script performed well so far.

When installing new Icinga2 for Windows Framework the AgentInstallWizard failes with
`Failed to connect to your Icinga Director at "https://192.168.100.47/director/". Please try again`
Somehow I doubt that, I tried with

- FQDN
- host name only
- IP
- with and without /director path
- with and without ending-slash behind "director"
- with and without https (as http should be redirected automatically

Icingaweb2 is running as a root application, so when navigating to director in icingaweb2 the path is as given.

The invocation is done with latest kickstart script as documented on the frameworks website.

**Comment by LordHepipud** [comment]
Can you please try to open a PowerShell and do the following:

```powershell
$Request = Invoke-WebRequest -Uri "https://192.168.100.47/director/" -UseBasicParsing;
$Request.StatusCode;
$Request.RawContent;
```

Which version of the Framework are you using?

**Comment by blindzero** [comment]
@LordHepipud thanks for putting that effort even on easter holidays ;)
Ok, it seems that it is some HTTPS issue...testing with the IP doesn't work out due to certificate issues:

```
Invoke-WebRequest : Die zugrunde liegende Verbindung wurde geschlossen: Für den geschützten SSL/TLS-Kanal konnte keine Vertrauensstellung hergestellt werden..
In Zeile:1 Zeichen:12
+ $Request = Invoke-WebRequest -Uri "https://192.168.100.47/director/"  ...
+            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebException
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand

```

When trying the DNS name the request works (Status 200) and brings this raw content:

```
HTTP/1.1 200 OK
Transfer-Encoding: chunked
Connection: keep-alive
Pragma: no-cache
Cache-Control: no-store, no-cache, must-revalidate
Content-Type: text/html; charset=UTF-8
Date: Mon, 13 Apr 2020 14:24:16 GMT
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Set-Cookie: Icingaweb2=u93hcddefgbvt33j3i4uvmbplc; path=/; secure; HttpOnly
Server: nginx/1.14.0 (Ubuntu)

<!DOCTYPE html>
<html class="no-js" lang="en">
<head>
  <meta charset="utf-8">
  <meta name="google" value="notranslate">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <meta http-equiv="cleartype" content="on">
  <title>Icinga Web 2 Login</title>
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, viewport-fi
t=cover">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="application-name" content="Icinga Web 2">
  <meta name="apple-mobile-web-app-title" content="Icinga">
  <link rel="mask-icon" href="/img/website-icon.svg" color="#0096BF">
  <base href="/">
  <script type="text/javascript">
  (function() {
    var html = document.getElementsByTagName('html')[0];
    html.className = html.className.replace(/no-js/, 'js');
  }());
  </script>
  <link rel="stylesheet" href="/css/icinga.min.css" media="all" type="text/css" />
  <link type="image/png" rel="shortcut icon" href="/img/favicon.png" />
  <link rel="apple-touch-icon" href="/img/touch-icon.png">
</head>
<body id="body" class="loading">
<pre id="responsive-debug"></pre>
<div id="layout" class="default-layout">
<div id="login">
    <div class="login-wrapper">
        <div class="login-form centered-ghost" data-base-target="layout">
            <div class="login-form-content centered-content">
                <div role="status" class="sr-only">
                    Welcome to Icinga Web 2. For users of the screen reader Jaws full and expectant compliant accessibility is possible only with use of the Firefox browser. VoiceOver on Mac OS X is tested on Chrome, Safari and Firefox.
            </div>
                <div class="logo-wrapper"><div id="icinga-logo" class="invert" aria-hidden="true"></div></div>
                                <form id="form_login-xfwouqzpnikt" name="form_login" enctype="application/x-www-form-urlencoded" class="icinga-controls" method="post" action="/authentication/login?redirect=director%2F">
<div class="control-group"><div class="control-label-group"><span id="form_login_username-xfwouqzpnikt-label"><label for="form_login_username-xfwouqzpnikt" class="control-label required">Username</label></span></div><input type="text" name="username" id="form_login_username-xfwouqzpnikt" value="" autocapitalize="off" class="autofocus" aria-required="true" required=""></div>
<div class="control-group"><div class="control-label-group"><span id="form_login_password-xfwouqzpnikt-label"><label for="form_login_password-xfwouqzpnikt" class="control-label required">Password</label></span></div><input type="password" name="password" id="form_login_password-xfwouqzpnikt" value="" class="" aria-required="true" required=""></div>

<input type="hidden" name="redirect" value="director/" id="form_login_redirect-xfwouqzpnikt">

<input type="hidden" name="formUID" value="form_login" id="form_login_formUID-xfwouqzpnikt">

<input type="hidden" name="CSRFToken" value="751301887|cec75c000a13f81aa676cf6f5ad50c28a5d7b4e31d8687178e106a07972f4117"  id="CSRFToken">
<div class="control-group form-controls">
<input type="submit" name="btn_submit" id="form_login_btn_submit-xfwouqzpnikt" value="Login" class="btn-primary" data-progress-label="Logging in"><div class="spinner "><i  aria-hidden="true" class="icon-spin6"></i></div></div></form>
         <div id="login-footer">
                    <p>Icinga Web 2 &copy; 2013-2020</p>
                    <a href="https://icinga.com/">icinga.com</a>                </div>
            </div>
        </div>
        <ul id="social">
            <li>
            <a href="https://twitter.com/icinga" target="_blank" title="Icinga on Twitter" aria-label="Icinga on Twitter
"><i  aria-hidden="true" class="icon-twitter"></i></a>            </li>
            <li>
                <a href="https://www.facebook.com/icinga" target="_blank" title="Icinga on Facebook" aria-label="Icinga on Facebook"><i  aria-hidden="true" class="icon-facebook-squared"></i></a>            </li>
            <li><a href="https://github.com/Icinga" target="_blank" title="Icinga on GitHub" aria-label="Icinga on GitHub"><i  aria-hidden="true" class="icon-github-circled"></i></a>            </li>
        </ul>
    </div>
</div>
</div>
<script type="text/javascript">
  (function() {
    if (document.defaultView && document.defaultView.getComputedStyle) {
      var matched;
      var html = document.getElementsByTagName('html')[0];
      var element = document.getElementById('layout');
      var name = document.defaultView
        .getComputedStyle(html)['font-family']
        .replace(/['",]/g, '');

      if (null !== (matched = name.match(/^([a-z]+)-layout$/))) {
        element.className = element.className.replace('default-layout', name);
        if ('object' === typeof window.console) {
          window.console.log('Icinga Web 2: setting initial layout to ' + name);
        }
      }
    }
  }());
</script>
<div id="collapsible-control-ghost" class="collapsible-control">
    <button type="button">
        <i  class="expand-icon icon-angle-double-down" role="img" title="Expand" aria-label="Expand"></i>        <i  class="collapse-icon icon-angle-double-up" role="img" title="Collapse" aria-label="Collapse"></i>    </button>
</div>
<script type="text/javascript" src="/js/icinga.min.js"></script>
<script type="text/javascript">
window.name = 'Icinga-xfwouqzpnikt';
var icinga = new Icinga({
  baseUrl: ''
});
</script>
</body>
</html>
```

**Comment by LordHepipud** [comment]
Okay so basically the call should be done by using the FQDN. Im still wondering why the error persists.
Could it be that your host is already registered within the Icinga Director?

You can check that on

```
Hosts -> <your host> -> Agent
```

If there is a button `Drop Self-Service Key`, then the host is already registered and there might be the issue. In addition, using the latest Version of the Framework (snapshot / 1.0.1) is mandatory, as previously an issue was present within the logic of the fetching.

Is this a new Windows System or have you been using the old module on this specific system before?

You are welcome by the way - happy easter :D

**Comment by blindzero** [comment]
It is quite strange at with #58 I could set it up by using the manual way and not the self-service endpoint. So in general I would assume that connections can work out (also because of several dozens of windows server agents running at the moment).

The button mentioned is not there but no matter for which host...
The windows system it self is not fresh, it was deployed with the old Icinga2-Windows-Kickstart-agent before.

So far I always selected stable. Trying now snapshot doesn't change anything.

**Comment by LordHepipud** [comment]
This is weird and I have no quick solution here. In general what should happen once the template API key is provided and the URL is correct, that the Director will fetch the data.

Can you please try this:

```powershell
Get-IcingaDirectorSelfServiceConfig -DirectorUrl <url to director> -ApiKey <the template API key>
```

Is this actually fetching any information?

To check if a local key is configured, you can also use this

```powershell
Show-IcingaDirecorSelfServiceKey
```

If no key is configured - you are fine. If a key is there, you can try this:

```powershell
Remove-IcingaDirectorSelfServiceKey
```

Afterwards please try again

**Comment by blindzero** [comment]
P.S. Unsure if that may be an issue: we changed the server's URL. So it was icinga2.INT.DOMAIN.TLD before. Because of remote access we changed it to monitoring.OTHERDOMAIN.TLD.
So we have to differentiate between icinga's certificate which is running at old hostname, and the webservers frontend certificate running the new one.
So far this wasn't an issue, the connection was made to icinga's 5665 (?) which was running with the old (internal) domain.

However, manual configuration works anyhow with the old domain. 
Using the new (right frontend SSL) domain doesn't work as the agent rejects the certificate because of hostname mismatch.

**Comment by LordHepipud** [comment]
Honestly I'm not sure right now if this is the issue. The module is having issues wit actually fetching data from the Director. It could only be a problem if the Director for some reason is unable to connect to the Icinga 2 core locally.

Are the Cmdlets printing anything useful?

**Comment by blindzero** [comment]
I think the new framework stack is confused by this hostname stuff...
`Get-IcingaDirectorSelfServiceConfig` show SSL handshake error when using the "old" FQDN which is used by Icinga's master certificate for the backend.
Using the "new" FQDN which is used by the fronend's webserver it's bringing up 

```
Icinga Director Self-Service has thrown an error: Got invalid API key "2bbbbee51acef0a7d4852478e428270b80c8f22e"
In C:\Program Files\WindowsPowerShell\Modules\icinga-powershell-framework\lib\apis\Get-IcingaDirectorSelfServiceConfig.psm1:29
Zeichen:9
+         throw 'Icinga Director Self-Service has thrown an error: ' +  ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Icinga Director...28270b80c8f22e":String) [], RuntimeException
    + FullyQualifiedErrorId : Icinga Director Self-Service has thrown an error: Got invalid API key "2bbbbee51acef0a7d4852478e428270b80c8f22e"
```

Funny, the agent is now configured manually and working with my master (all services back to OK state), but
`C:\Windows\system32> Show-IcingaDirecorSelfServiceKey` show up

```
There is no Self-Service Api key configured on this system
```

**Comment by LordHepipud** [comment]
Basically im not worrying about the SSL handshake error, because this is resolvable and using the "new" FQDN would be the way to go. The other configuration is pushed by the Director anyway and as long as this configuration is correct, the Agent will be configured properly.

The API Key however is not valid. Can you please re-check with the Host-Template you wish to use for the Windows Systems is the keys are matching properly or if they changed?

This error here mainly is caused by either not using an "Agent" Host-Template or the Template Key itself being wrong.

**Comment by blindzero** [comment]
The API key seems to be right as this exactly is working on the manual way...?

**Comment by LordHepipud** [comment]
What do you mean by manual way? The manual way is in general not using a self service API key, but the ticket for the certificate signing.

The API-Key is not equal to the Icinga 2 ticket.

**Comment by blindzero** [comment]
@LordHepipud doh! ok...sorry...didn't made that up yet...now I got it...unfortunately the documentation at https://icinga.com/docs/director/latest/doc/74-Self-Service-API/ is only really half the way...it is completely missing all about this template, API Key stuff...

However, I now followed this quite right (I guess) and get the error 

```
Please enter your Self-Service API key: c7d4e603e20dc75ebecd9912aac415759123b3e5
Failed to register host within Icinga Director. Full error: "The term 'Get-NetRoute' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.". Please re-enter your SelfService API Key. If this prompt continues ensure you are using an Agent template or drop your host key at "Hosts -> acmp.corp.DOMAIN.TLD -> Agent" (Default: "c7d4e603e20dc75ebecd9912aac415759123b3e5"):
```

- Get-NetRoute is really not existing, I am running on Windows Server 2008 R2 here with Powershell 5
```
PS C:\Windows\system32> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      5.1.14409.1018
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14409.1018
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```
- I am 100% sure that the host I want to register isn't even existing

Still something on my side or really an issue within the new framework?

**Comment by LordHepipud** [comment]
I will try to get the documentation optimized there - thank you very much.

That one is used to determine which interface should be used to accept/connect to Icinga 2. The reason for that is, that the Self-Service API will register the Host within the Director and assign a host name and the address to it.

Get-NetRoute is a general Cmdlet available in Windows, but if it is missing in Windows 2008 R2, then we have a slight Problem. The problem is that we do not support Windows 2008 R2 out of the box.

Im not sure yet if there is a equivalent available in 2008 R2 which could be used to get the same amount of data

**Comment by iiM-Dawnbreaker** [comment]
Since we are still not capable of detaching every single 2008 R2 Server I'm also interested in a way to make the module working for 2008 R2. :)

**Comment by LordHepipud** [comment]
I just created a new branch to test a first solution attempt for this. could you please test and verify if this resolves your issue?

I also added some testing details allowing you to verify if the correct interface is detected

**Comment by blindzero** [comment]
Hi @LordHepipud 
sorry for being absent so long. I wanted to have a look and test but couldn't find a defined branch. In support/1.1.0 I couldn't find any commit into that?

**Comment by LordHepipud** [comment]
Hello, sorry for the late reply.
The feature is developed in an own branch: [feature/network_routes](https://github.com/Icinga/icinga-powershell-framework/pull/60)

**Comment by LordHepipud** [comment]
Fixed with #60
